Extract workflow from purged history does not work

xref https://github.com/galaxyproject/galaxy/issues/18113#issuecomment-2144827835

Trying to extract a workflow out of a purged history gives a 500 error.

**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 24.1
Commit: (run `git rev-parse HEAD` if you run this Galaxy server)

**Browser and Operating System**
Operating System: Windows, Linux, macOS
Browser: Firefox, Chrome, Chrome-based, Safari

**To Reproduce**
Steps to reproduce the behavior:
1. Purge a history
2. Switch to that purged history from the histories list
3. Extract a workflow from the history
4. See error

https://github.com/user-attachments/assets/bd6dd1c1-7a33-45b8-ab18-4d08233e0fdb


**Additional context**
 Stack trace:
```
galaxy.workflow.extract WARNING 2025-02-03 16:23:05,153 [pN:main.1,p:312402,tN:WSGI_1] job_id 4 not found in job_id2representative_job {}
galaxy.web.framework.middleware.error ERROR 2025-02-03 16:23:05,154 [pN:main.1,p:312402,tN:WSGI_1] Uncaught Exception
Traceback (most recent call last):
  File "/home/dlopez/dev/galaxy/lib/galaxy/web/framework/middleware/error.py", line 167, in __call__
    app_iter = self.application(environ, sr_checker)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/.venv/lib/python3.12/site-packages/paste/httpexceptions.py", line 635, in __call__
    return self.application(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/web/framework/base.py", line 176, in __call__
    return self.handle_request(request_id, path_info, environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/web/framework/base.py", line 271, in handle_request
    body = method(trans, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/webapps/galaxy/controllers/workflow.py", line 310, in build_from_current_history
    stored_workflow = extract_workflow(
                      ^^^^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/workflow/extract.py", line 45, in extract_workflow
    steps = extract_steps(
            ^^^^^^^^^^^^^^
  File "/home/dlopez/dev/galaxy/lib/galaxy/workflow/extract.py", line 138, in extract_steps
    raise AssertionError("Attempt to create workflow with job not connected to current history")
AssertionError: Attempt to create workflow with job not connected to current history

```

--- a/lib/galaxy/model/__init__.py
+++ b/lib/galaxy/model/__init__.py
@@ -3645,6 +3645,10 @@ class History(Base, HasTags, Dictifiable, UsesAnnotations, HasName, Serializable
         """Return all active contents ordered by hid."""
         return self.contents_iter(types=["dataset", "dataset_collection"], deleted=False, visible=True)
 
+    @property
+    def visible_contents(self):
+        return self.contents_iter(types=["dataset", "dataset_collection"], visible=True)
+
     def contents_iter(self, **kwds):
         """
         Fetch filtered list of contents of history.

--- a/lib/galaxy/workflow/extract.py
+++ b/lib/galaxy/workflow/extract.py
@@ -283,7 +283,7 @@ class WorkflowSummary:
         # just grab the implicitly mapped jobs and handle in second pass. Second pass is
         # needed because cannot allow selection of individual datasets from an implicit
         # mapping during extraction - you get the collection or nothing.
-        for content in self.history.active_contents:
+        for content in self.history.visible_contents:
             self.__summarize_content(content)
 
     def __summarize_content(self, content):

--- a/templates/webapps/galaxy/workflow/build_from_current_history.mako
+++ b/templates/webapps/galaxy/workflow/build_from_current_history.mako
@@ -112,6 +112,7 @@ into a workflow will be shown in gray.</p>
     <%
     cls = "toolForm"
     tool_name = "Unknown"
+    checked_job = "checked" if any(True for d in datasets if not d[1].deleted) else ""
     if hasattr( job, 'is_fake' ) and job.is_fake:
         cls += " toolFormDisabled"
         disabled = True
@@ -142,7 +143,10 @@ into a workflow will be shown in gray.</p>
                     %if disabled:
                         <div style="font-style: italic; color: gray">${disabled_why}</div>
                     %else:
-                        <div><input type="checkbox" name="job_ids" value="${job.id}" checked="true" />Include "${tool_name}" in workflow</div>
+                        <div><input type="checkbox" name="job_ids" value="${job.id}" ${checked_job} />Include "${tool_name}" in workflow</div>
+                        %if not checked_job:
+                            ${ render_msg( "All job outputs have been deleted", status="info" ) }
+                        %endif
                         %if tool_version_warning:
                             ${ render_msg( tool_version_warning, status="warning" ) }
                         %endif

--- a/templates/webapps/galaxy/workflow/build_from_current_history.mako
+++ b/templates/webapps/galaxy/workflow/build_from_current_history.mako
@@ -88,7 +88,7 @@ into a workflow will be shown in gray.</p>
     <div class="warningmark">${warning}</div>
 %endfor
 
-<form method="post" action="${h.url_for(controller='workflow', action='build_from_current_history')}">
+<form method="post" action="${h.url_for(controller='workflow', action='build_from_current_history', history_id=trans.security.encode_id(history.id))}">
 <div class='form-row'>
     <label>Workflow name</label>
     <input name="workflow_name" type="text" value="Workflow constructed from history '${ util.unicodify( history.name )}'" size="60"/>

--- a/test/unit/workflows/test_extract_summary.py
+++ b/test/unit/workflows/test_extract_summary.py
@@ -121,6 +121,10 @@ class MockHistory:
     def active_contents(self):
         return self.active_datasets
 
+    @property
+    def visible_contents(self):
+        return self.active_contents
+
 
 class MockTrans:
     def __init__(self, history):

